### PR TITLE
38 fix types in the ast

### PR DESCRIPTION
**Pull request by pablolh**
Changes

- Implement ControlledSemantic to be able to express CNOT as a
  controlled X, for instance
- Separate validation of qubit ranges from type-checking in a separate
  ANTLR tree visitor
- snake_case instead of camelCase for variable names
- Add docstrings to parsing classes
- The GateLibrary superclass is used to factorize code for gateset-aware
classes (e.g. TypeChecker)
- Semantic classes are removed, absorbed into Gate subclasses
- Move ALL ANTLR parsing classes to parsing folder - this has already
  been done, but only partially
- Gates are specified directly with Python functions and a decorator
- Lay the basis of handling of anonymous gates
- Writer turned into impl pattern
- Rename AST -> IR
- Make test_testinterpreter a proper unit test, before it was testing
  parsing too
- Visitor pattern for the SquirrelIR class
- Use those visitors in e.g. writer and test_interpreter
- Various code fixes
- Use inspect to know Python function signatures
 This makes the gate semantic API nice to use from Python.

### CODE REVIEW

File: opensquirrel/default_gates.py [hunk 1]
```diff
@@ -1,104 +1,74 @@
...
+    return ControlledGate(control, x(target))
+
+
+@named_gate
+def cz(control: Qubit, target: Qubit) -> Gate:
+    return ControlledGate(control, z(target))
+
+
+@named_gate
+def cr(control: Qubit, target: Qubit, theta: Float) -> Gate:
```
Comment (elenbaasc) [inline opensquirrel/default_gates.py:67]: `cr` by default is Controlled-Rz?

File: opensquirrel/mckay_decomposer.py [hunk 2]
```diff
@@ -3,135 +3,115 @@
 
 import numpy as np
 
-from opensquirrel.common import ATOL, ArgType
-from opensquirrel.gates import SingleQubitAxisAngleSemantic, queryEntry, querySemantic, querySignature
-from opensquirrel.squirrel_ast import SquirrelAST
+from opensquirrel.common import ATOL, normalize_angle
+from opensquirrel.default_gates import rz, x90
+from opensquirrel.squirrel_ir import BlochSphereRotation, Float, Gate, Int, Qubit, SquirrelIR, Statement
```
Comment (elenbaasc) [inline opensquirrel/mckay_decomposer.py:8]:
```suggestion
from opensquirrel.squirrel_ir import BlochSphereRotation, Float, Gate, Qubit, SquirrelIR
```

File: opensquirrel/mckay_decomposer.py [hunk 3]
```diff
@@ -3,135 +3,115 @@
...
+        combined_angle = 2 * acos(cos(a / 2) * cos(b / 2) - sin(a / 2) * sin(b / 2) * np.dot(l, m))
 
-        if abs(sin(combinedAngle / 2)) < ATOL:
-            self.oneQubitGates.pop(qubit)
+        if abs(sin(combined_angle / 2)) < ATOL:
+            self.accumulated_1q_gates.pop(qubit)
             return
 
-        combinedAxis = (
+        combined_axis = (
```
Comment (elenbaasc) [inline opensquirrel/mckay_decomposer.py:81]: PyCharm complains that this piece of code is unreachable. Is the above if statement always true?

File: opensquirrel/mckay_decomposer.py [hunk 4]
```diff
@@ -3,135 +3,115 @@
...
-            self._processSingleOperation(operation)
+        if len(qubit_arguments) >= 2:
+            [self.flush(q) for q in qubit_arguments]
+            self.output.add_gate(gate)
+            return
 
-        self._flush_all()
+        if len(qubit_arguments) == 0:
+            assert False, "Unsupported"
+            return
```
Comment (elenbaasc) [inline opensquirrel/mckay_decomposer.py:99]: PyCharm complains that the `return` statement is unreachable, does the code exit when you have `assert False`?

File: opensquirrel/writer.py [hunk 5]
```diff
@@ -1,41 +1,37 @@
-from opensquirrel.common import ArgType
-from opensquirrel.gates import querySignature
+from opensquirrel.squirrel_ir import Comment, Float, Gate, Int, Qubit, SquirrelIR, SquirrelIRVisitor, Statement
```
Comment (elenbaasc) [inline opensquirrel/writer.py:1]:
```suggestion
from opensquirrel.squirrel_ir import Comment, Float, Gate, Int, Qubit, SquirrelIR, SquirrelIRVisitor
```

File: opensquirrel/writer.py [hunk 6]
```diff
@@ -1,41 +1,37 @@
...
-            return f"{squirrelAST.qubitRegisterName}[{arg}]"
-        if t == ArgType.INT:
-            return f"{int(arg)}"
-        if t == ArgType.FLOAT:
-            return f"{float(arg):.{Writer.NUMBER_OF_SIGNIFICANT_DIGITS}}"
+    def visit_qubit(self, qubit: Qubit):
+        return f"{self.qubit_register_name}[{qubit.index}]"
 
-        assert False, "Unknown argument type"
+    def visit_int(self, _int: Int):
```
Comment (elenbaasc) [inline opensquirrel/writer.py:14]: For consistency you could use `i` here as well, iso `_int`, as done in the `squirrel_ir.py` module. (same for `_float`).

File: opensquirrel/writer.py [hunk 7]
```diff
@@ -1,41 +1,37 @@
...
+    def visit_float(self, _float: Float):
+        return f"{_float.value:.{self.number_of_significant_digits}}"
 
-        for operation in squirrelAST.operations:
-            if isinstance(operation, str):
-                comment = operation
-                assert "*/" not in comment, "Comment contains illegal characters"
+    def visit_gate(self, gate: Gate):
+        if gate.is_anonymous():
+            self.output += "<anonymous gate>\n"
```
Comment (elenbaasc) [inline opensquirrel/writer.py:22]: Add a hyphen?
```suggestion
            self.output += "<anonymous-gate>\n"
```

File: test/test_replacer.py [hunk 8]
```diff
@@ -1,32 +1,31 @@
 import unittest
 
-from opensquirrel.default_gates import DefaultGates
-from opensquirrel.replacer import Replacer
-from opensquirrel.squirrel_ast import SquirrelAST
+from opensquirrel import replacer
+from opensquirrel.default_gates import *
+from opensquirrel.squirrel_ir import Gate, Qubit, SquirrelIR
```
Comment (elenbaasc) [inline test/test_replacer.py:5]:
```suggestion
from opensquirrel.squirrel_ir import Qubit, SquirrelIR
```

File: test/test_testinterpreter.py
Comment (elenbaasc) [inline test/test_testinterpreter.py:1]: Rename in light of renaming the module that is being tested here.

**Review by elenbaasc — COMMENTED** [review]
I have been through the code, but mostly to check if there are no weird things or things going wrong. I haven't been able to check/optimize the functionalities. It's quite a big refactor and a really nice one as well! I would still need some time to fully understand how everything links together, but our meeting this morning helped a lot.

The test coverage is now 82%, it would be good if we could increase this (perhaps this could be an issue of its own though).

**Comment by pablolh** [comment]
Thanks Chris for the review, will look asap

File: opensquirrel/squirrel_ir.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
+import inspect
+from abc import ABC
+from dataclasses import dataclass
+from functools import wraps
+from typing import Callable, List, Optional, Tuple, Union
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/squirrel_ir.py:5]: Argh
I was hoping isort would do it for me

File: docs/reference.md [hunk 10]
```diff
@@ -14,9 +14,9 @@
 
 ::: opensquirrel.replacer
 
-::: opensquirrel.squirrel_ast
+::: opensquirrel.squirrel_ir
 
-::: opensquirrel.squirrel_ast_creator
+::: opensquirrel.squirrel_ir_creator
```
[the remark this answers is not in the document]
Comment (pablolh) [reply docs/reference.md:19]: So we is the goal to not document things in parsing/ or utils?

File: docs/reference.md [hunk 11]
```diff
@@ -14,9 +14,9 @@
```
[the remark this answers is not in the document]
Comment (pablolh) [reply docs/reference.md:14]: yes

**Comment by pablolh** [comment]
I've tried increasing the test coverage a bit, by looking at what was not covered
Most of the non-covered code is now **generated code from ANTLR**... Not sure if I can do much about it, the `test_parsing.py` is already pretty complete

File: opensquirrel/squirrel_ir.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/squirrel_ir.py]: It is weird indeed that isort didn't pick this up....

File: docs/reference.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (elenbaasc) [reply docs/reference.md]: No, they should also be included in the docs.

**Comment by elenbaasc** [comment]
> I've tried increasing the test coverage a bit, by looking at what was not covered Most of the non-covered code is now **generated code from ANTLR**... Not sure if I can do much about it, the `test_parsing.py` is already pretty complete

Yes, I see, good point. We could exclude the generated files by adding them to the `omit` list of the coverage tool in the `pyproject.toml`:
```toml
[tool.coverage.run]
branch = true
omit = ["tests/*", "**/.tox/**","opensquirrel/parsing/antlr/*"]
```

File: docs/reference.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (pablolh) [reply docs/reference.md]: I've automated this using a variant of this https://mkdocstrings.github.io/recipes/

File: scripts/gen_reference_page.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+repo = Path(__file__).parent.parent
+opensquirrel = repo / "opensquirrel"
+reference_file = repo / "docs" / "reference.md"
+
+reference_md_content = "<!--- This file was automatically generated by scripts/gen_reference_page.py --->\n\n"
+for path in sorted(opensquirrel.rglob("*.py")):
+    module_path = path.relative_to(repo).with_suffix("")
+
+    print(module_path)
```
Comment (elenbaasc) [inline scripts/gen_reference_page.py:17]: Do we want to introduce `logging` for this, in stead of print statements? If so, we can add a TODO in the places and create an issue to pick this up later.

File: test/test_mckay_decomposer.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
+import unittest
+
+from opensquirrel import circuit_matrix_calculator, mckay_decomposer
+from opensquirrel.default_gates import *
+from opensquirrel.squirrel_ir import Float, Qubit, SquirrelIR
+
+
+def are_matrices_equal_up_to_global_phase(matrix_a, matrix_b):
```
Comment (elenbaasc) [inline test/test_mckay_decomposer.py:8]:
```suggestion
def check_equivalence_up_to_global_phase(matrix_a, matrix_b):
```

File: test/test_mckay_decomposer.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    phase_difference = matrix_a[first_non_zero] / matrix_b[first_non_zero]
+
+    return np.allclose(matrix_a, phase_difference * matrix_b)
+
+
+class DecomposeMcKayTests(unittest.TestCase):
+    def checkMcKayDecomposition(self, squirrel_ir, expected_ir=None):
+        """
+        Check whether the mcKay decomposition transformation applied to the input IR preserves the
+        circuit matrix up to an irrelevant global phase factor.
```
Comment (elenbaasc) [inline test/test_mckay_decomposer.py:25]:
```suggestion
        circuit matrix up to a global phase factor.
```

File: test/test_mckay_decomposer.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    if abs(matrix_b[first_non_zero]) < ATOL:
+        return False
+
+    phase_difference = matrix_a[first_non_zero] / matrix_b[first_non_zero]
+
+    return np.allclose(matrix_a, phase_difference * matrix_b)
+
+
+class DecomposeMcKayTests(unittest.TestCase):
+    def checkMcKayDecomposition(self, squirrel_ir, expected_ir=None):
```
Comment (elenbaasc) [inline test/test_mckay_decomposer.py:22]:
```suggestion
    def check_mckay_decomposition(self, squirrel_ir, expected_ir=None):
```

File: opensquirrel/parsing/antlr/squirrel_ir_creator.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        GateLibrary.__init__(self, gate_set, gate_aliases)
+        self.squirrel_ir = None
+
+    def visitProg(self, ctx):
+        number_of_qubits, qubit_register_name = self.visit(ctx.qubitRegisterDeclaration())
+
+        self.squirrel_ir = SquirrelIR(number_of_qubits=number_of_qubits, qubit_register_name=qubit_register_name)
+
+        for gate_application in ctx.gateApplication():
+            self.visit(gate_application)
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/parsing/antlr/squirrel_ir_creator.py:26]: I think `gate_application` is the right name, so I've fixed the two others

File: opensquirrel/squirrel_ir.py [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+from dataclasses import dataclass
+from functools import wraps
+from typing import Callable, List, Optional, Tuple
+
+import numpy as np
+
+from opensquirrel.common import ATOL, normalize_angle
+
+
+class SquirrelIRVisitor(ABC):
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/squirrel_ir.py:12]: I guess this interface is just here to make the typing annotations work
It's equivalent to the base visitor class in tree-gen, for instance
A (somewhat complicated) explanation of the visitor pattern: https://en.wikipedia.org/wiki/Visitor_pattern

File: scripts/gen_reference_page.py (hunk 12 above)
Comment (pablolh) [reply scripts/gen_reference_page.py:17]: This `print` was left accidentally, I didn't intend to leave it
now removed
Nevertheless, logging can be nice, so we can create the issue.

File: test/test_mckay_decomposer.py (hunk 13 above)
Comment (pablolh) [reply test/test_mckay_decomposer.py:8]: For me if I rename it that way, the function should return nothing and throw when matrices are not equal up to global phase
So
`check` -> assertion (function throws for instance)
`is` or `are` -> return a boolean, the `False` return value can also be useful in some context


I've renamed it as you suggest but then made it do the `self.assertTrue` on top and return void.

File: opensquirrel/squirrel_ir.py [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    def accept(self, visitor: SquirrelIRVisitor):
+        return visitor.visit_int(self)
+
+
+@dataclass
+class Qubit(Expression):
+    index: int
+
+    def __hash__(self):
+        return hash(self.index)
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/squirrel_ir.py:68]: I blindly followed this https://docs.python.org/3.5/reference/datamodel.html#object.__hash__

In this case the qubit index alone would work perfectly I believe
But hashing functions can be tricky, some (pseudo)randomization (done by `hash()`) can be useful to get balanced hash tables, etc...

File: opensquirrel/squirrel_ir.py [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+    generator: Optional[Callable[..., "BlochSphereRotation"]] = None
+
+    def __init__(self, qubit: Qubit, axis: Tuple[float, float, float], angle: float, phase: float = 0):
+        self.qubit: Qubit = qubit
+        self.axis = self._normalize_axis(np.array(axis).astype(np.float64))
+        self.angle = normalize_angle(angle)
+        self.phase = normalize_angle(phase)
+
+    @staticmethod
+    def _normalize_axis(axis):
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/squirrel_ir.py:100]: yes

File: opensquirrel/squirrel_ir.py [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+class Gate(Statement, ABC):
+    arguments: Optional[Tuple[Expression, ...]] = None
+    generator: Optional[Callable[..., "Gate"]] = None
+
+    @property
+    def name(self) -> Optional[str]:
+        return self.generator.__name__ if self.generator else None
+
+    def is_anonymous(self) -> bool:
+        return self.arguments is None
```
[the remark this answers is not in the document]
Comment (pablolh) [reply opensquirrel/squirrel_ir.py:87]: yes
tbh I'm not super fond of @property but it can be convenient sometimes

**Review by elenbaasc — APPROVED** [review]

**Pull request merged**